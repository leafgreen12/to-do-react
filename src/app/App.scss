.todos {
  width: 500px;
  margin: 0 auto;

  h1 {
    margin: 0;
  }
  h2 {
    margin: 0 0 50px 0;
    color: rgba(0,0,0,.4);
  }
}
