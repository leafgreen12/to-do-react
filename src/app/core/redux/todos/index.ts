export * from './todos.states';
export * from './todos.action';
export * from './todos.reducer';
