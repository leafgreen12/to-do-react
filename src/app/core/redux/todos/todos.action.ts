import {TodoModel} from "../../interfaces/todo.interface";

export enum ETodos {
    ADD_TODO = '[TODOS] add todo',
    EDIT_TODO = '[TODOS] edit todo',
    DELETE_TODO = '[TODOS] delete todo',
}

export class AddTodo {
    public readonly type = ETodos.ADD_TODO;
    constructor(
        public data: TodoModel) {}
}

export class EditTodo {
    public readonly type = ETodos.EDIT_TODO;
    constructor(
        public data: TodoModel) {}
}

export class DeleteTodo {
    public readonly type = ETodos.DELETE_TODO;
    constructor(
        public idTodo: number) {}
}

export type HandleTodoActions =
    AddTodo |
    DeleteTodo |
    EditTodo;
