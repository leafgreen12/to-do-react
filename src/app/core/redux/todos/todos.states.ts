import {TodoModel} from "../../interfaces/todo.interface";

export interface ITodoState {
    items: TodoModel[]
    dispatch?: any
}
