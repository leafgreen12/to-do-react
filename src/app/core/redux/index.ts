import * as fromTodos from './todos';

export interface IAppState {
    todos: fromTodos.ITodoState;
}
