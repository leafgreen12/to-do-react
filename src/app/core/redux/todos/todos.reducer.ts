import {ITodoState} from './todos.states';
import {ETodos, HandleTodoActions} from "./todos.action";
import {TodoModel} from "../../interfaces/todo.interface";

export const initialState = {} as ITodoState

const todos = (state = initialState, action: HandleTodoActions) => {
    switch (action.type) {
        case ETodos.ADD_TODO:
            const newItem = {...state} as ITodoState;
            action.data.id = newItem.items.length + 1;
            newItem.items.push(action.data as TodoModel);
            return newItem;
        case ETodos.EDIT_TODO:
            const newEditItem = {...state} as ITodoState;
            const existItem = newEditItem.items.find(item => item.id === action.data.id);
            if (existItem) {
                existItem.title = action.data.title;
            }
            return newEditItem;
        case ETodos.DELETE_TODO:
            const newArray = {...state} as ITodoState;
            newArray.items = newArray.items.filter(item => item.id !== action.idTodo)
            return newArray;
        default:
            return state
    }
}

export default todos
