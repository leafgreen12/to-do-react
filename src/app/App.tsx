import React, {Component} from 'react';
import './App.scss';
import { Provider } from "react-redux";
import TodoAddItem from "./todo-add-item/todo-add-item";
import {TodosService} from "./core/services/todos.service";
import TodosLists from "./todos-lists/todos-lists";
import todos from "./core/redux/todos/todos.reducer";
import {HandleTodoActions, ITodoState} from "./core/redux/todos";
import {TodoModel} from "./core/interfaces/todo.interface";
import store from "./core/redux/redux-form/store";
export const TodosContext = React.createContext({
    items: [],
    dispatch: () => null
} as Readonly<ITodoState>);

class App extends Component<any, any> {
    constructor(props: any) {
        super(props);
        this.state = {
            items: [] as TodoModel[],
            dispatch: (action: any) => this.reducer(action)
        } as Readonly<ITodoState>
    }
    componentWillUnmount() {
        this.getData();
    }

    async getData() {
        const todosService = new TodosService();
        const data = await todosService.getTodos();
        const stateApp = {
            items: data as TodoModel[],
            dispatch: (action: any) => this.reducer(action)
        }
        this.setState(stateApp);
    }

    reducer(action: HandleTodoActions) {
        let state = {...this.state} as ITodoState;
        this.setState(todos(state, action));
    }

    render() {
        return (
            <div className="todos">
                <h1>Todo's App</h1>
                <Provider store={store}>
                    <TodosContext.Provider value={this.state}>
                        <TodoAddItem dispatch={(action: any) => this.reducer(action)} onSubmit={(action: any) => this.reducer(action)}/>
                        <TodosLists/>
                    </TodosContext.Provider>
                </Provider>
            </div>
        );
    }
}

export default App;
