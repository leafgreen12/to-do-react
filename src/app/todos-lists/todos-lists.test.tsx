import React, {useContext} from 'react';
import {act, render, screen} from '@testing-library/react';
import {unmountComponentAtNode} from 'react-dom';
import Enzyme, {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16';
import {Provider} from "react-redux";
import store from "./../core/redux/redux-form/store";
import * as fromTodosAction from "./../core/redux/todos";
import TodosLists from "./todos-lists";
import {TodosService} from "../core/services/todos.service";
import App, {TodosContext} from "../App";

// @ts-ignore
(Enzyme as any).configure({adapter: new Adapter()})
describe('TodoList', () => {
    let container = null as any;
    let todosLists = null as any;
    beforeEach(() => {
        // cài đặt một DOM element như là target cho render
        container = document.createElement("div");
        document.body.appendChild(container);
        todosLists = shallow(<TodosLists/>)
    });

    afterEach(() => {
        // dọn dẹp lúc thoát
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    it('should render todosLists correctly', () => expect(todosLists).toMatchSnapshot());
});
