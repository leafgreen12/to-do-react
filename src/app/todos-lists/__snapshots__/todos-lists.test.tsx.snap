// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render todosLists correctly 1`] = `ShallowWrapper {}`;

exports[`TodoList should render todosLists correctly 1`] = `ShallowWrapper {}`;
